### DISCUSSION TITLE
Excessive re-rendering from useForm hook

### DISCUSSION THREAD

**Discussion by rfdrake27** [mantinedev/mantine#1193] [Feature requests]
Hello, I'm noticing that updating any value within the form will trigger a re-render of the parent component where the useForm hook is located. Even after destructuring the returned object from useForm, I'm noticing that functions such as getInputProps or addListItem are also being changed. 

Are there plans to optimize performance of this hook? Such as wrapping functions in useCallback, or using refs?

Thank you Vitaly for the excellent library.

**Comment by rtivital** [comment]
I'll see what we can do, but that is not a priority.

**Reply by Arjan-Zuidema** [reply]
I think performance should be a priority? Rerenders can also cause a lot of DX frustration

**Comment by kaptinlin** [comment]
It re-render the whole form when interaction with form field

**Comment by thomasbeckford** [comment]
Im using <ColorInput /> component inside a form, and its re-rendering everytime I slide the mouse over the color grid.. I think its better to useRef in this case 😕 , right?

**Comment by SadRobinson** [comment]
Is there any solution to this problem?

**Reply by piyushsaha** [reply]
Hi, have you found any solution?

**Comment by BahaaZidan** [comment]
We had to move to https://github.com/react-hook-form/react-hook-form

**Reply by piyushsaha** [reply]
Hi. Is the performance better now?

**Reply by BahaaZidan** [reply]
With `react-hook-form`, it doesn't render the entire form when the user changes the value of one field. So yes, much better than Mantine.

**Reply by piyushsaha** [reply]
Are you using the Mantine input components?

I tried both `useFormik` hook from `formik` and the `useForm` hook from `@mantine/form` and in mid tier the performace is slow. Each keystroke is lagging.

**Reply by BahaaZidan** [reply]
Yes, I'm using `react-hook-form` with Mantine input components.

**Reply by piyushsaha** [reply]
Have you tried it with Date or Select components? I migrated to `react-hook-form` and I am having revalidatrion issues with these components. The `revalidateMode` is `onChange` and yet the error stays until submit is executed again.
Have you faced this?

**Reply by BahaaZidan** [reply]
no

**Comment by rcronin** [comment]
We are seeing this still an issue in Mantine 7.5.1. @rtivital - any guidance on how to limit the re-renders while using the useForm hook?

**Reply by rtivital** [reply]
This is expected behavior of controlled forms. If you want your form to be uncontrolled, use react-hook-form.

**Reply by gcsandesh** [reply]
@rtivital  What about using mantine form with {mode: 'uncontrolled'} ? Will it limit the re-renders?

**Reply by rtivital** [reply]
yes

**Comment by rtivital** [comment]
All performance issues have been resolved with uncontrolled form mode – https://mantine.dev/form/uncontrolled/
